File: connect6s/game.py
import numpy as np

BOARD_SIZE = 15
WIN_LENGTH  = 6
ACTION_SIZE = BOARD_SIZE * BOARD_SIZE * 2  # regular + strong moves

# Board cell values
EMPTY     =  0
BLACK_REG =  1
WHITE_REG = -1
BLACK_STR =  2
WHITE_STR = -2

# Strong piece award schedule: before moves 6, 13, 20, 27... (1-indexed)
# After move_count moves played, next move is move_count+1 (1-indexed)
# Award when next >= 6 and (next - 6) % 7 == 0

# ── Zobrist tables (module-level, built once) ─────────────────────────────────
# Piece types (non-empty): WHITE_STR=-2→0, WHITE_REG=-1→1, BLACK_REG=1→2, BLACK_STR=2→3
_PIECE_IDX = {-2: 0, -1: 1, 1: 2, 2: 3}

_rng = np.random.default_rng(0xDEADBEEF)
_CELL_TABLE   = _rng.integers(1, np.iinfo(np.uint64).max,
                               size=(4, BOARD_SIZE, BOARD_SIZE),
                               dtype=np.uint64)
_PLAYER_TABLE = _rng.integers(1, np.iinfo(np.uint64).max, dtype=np.uint64)
# strong count per player (0 or 1): shape [player_idx=0/1][count=0/1]
_STRONG_TABLE = _rng.integers(1, np.iinfo(np.uint64).max,
                               size=(2, 2), dtype=np.uint64)

# Initial hash for empty board, Black to move, both strong=0
_INIT_HASH = np.uint64(_STRONG_TABLE[0, 0]) ^ np.uint64(_STRONG_TABLE[1, 0])


class GameState:
    def __init__(self):
        self.board          = np.zeros((BOARD_SIZE, BOARD_SIZE), dtype=np.int8)
        self.current_player = 1           # 1=Black, -1=White
        self.move_count     = 0
        self.strong_pieces  = {1: 0, -1: 0}
        self.game_over      = False
        self.winner         = None        # 1, -1, or 0 (draw)
        self.zobrist_hash   = int(_INIT_HASH)

    # ── Copy ──────────────────────────────────────────────────────────────────
    def copy(self):
        s = GameState.__new__(GameState)
        s.board          = self.board.copy()
        s.current_player = self.current_player
        s.move_count     = self.move_count
        s.strong_pieces  = self.strong_pieces.copy()
        s.game_over      = self.game_over
        s.winner         = self.winner
        s.zobrist_hash   = self.zobrist_hash
        return s

    # ── Legal moves ───────────────────────────────────────────────────────────
    def get_legal_moves(self):
        """Return list of (row, col, is_strong)."""
        p          = self.current_player
        has_strong = self.strong_pieces[p] > 0
        moves      = []
        for r in range(BOARD_SIZE):
            for c in range(BOARD_SIZE):
                cell = int(self.board[r, c])
                if cell == EMPTY:
                    moves.append((r, c, False))
                    if has_strong:
                        moves.append((r, c, True))
                elif has_strong and abs(cell) == 1:
                    moves.append((r, c, True))
        return moves

    def get_legal_move_set(self):
        return set(self.get_legal_moves())

    # ── Action encoding ───────────────────────────────────────────────────────
    def action_to_index(self, row, col, is_strong):
        flat = row * BOARD_SIZE + col
        return flat + (BOARD_SIZE * BOARD_SIZE if is_strong else 0)

    def index_to_action(self, idx):
        base = BOARD_SIZE * BOARD_SIZE
        if idx >= base:
            return (idx - base) // BOARD_SIZE, (idx - base) % BOARD_SIZE, True
        return idx // BOARD_SIZE, idx % BOARD_SIZE, False

    # ── State transition ──────────────────────────────────────────────────────
    def make_move(self, row, col, is_strong=False):
        """Return new GameState (immutable-style). Updates Zobrist hash incrementally."""
        s = self.copy()
        p    = s.current_player
        pidx = 0 if p == 1 else 1

        old_cell  = int(s.board[row, col])
        new_cell  = 2 * p if is_strong else p
        h         = s.zobrist_hash

        # --- Remove old piece from hash ---
        if old_cell != EMPTY:
            h ^= int(_CELL_TABLE[_PIECE_IDX[old_cell], row, col])

        # --- Place piece ---
        if is_strong:
            assert s.strong_pieces[p] > 0,          "no strong piece"
            assert abs(old_cell) <= 1,               "strong move must target empty or regular piece"
            # XOR out old strong count for this player
            h ^= int(_STRONG_TABLE[pidx, s.strong_pieces[p]])
            s.strong_pieces[p] -= 1
            h ^= int(_STRONG_TABLE[pidx, s.strong_pieces[p]])
        else:
            assert old_cell == EMPTY, "cell not empty"

        s.board[row, col] = np.int8(new_cell)
        h ^= int(_CELL_TABLE[_PIECE_IDX[new_cell], row, col])

        s.move_count += 1

        if s._check_win(row, col, p):
            s.game_over = True
            s.winner    = p
        elif not np.any(s.board == EMPTY):
            s.game_over = True
            s.winner    = 0

        # Toggle player
        s.current_player = -p
        h ^= int(_PLAYER_TABLE)

        # Award strong pieces if scheduled
        nxt = s.move_count + 1
        if nxt >= 6 and (nxt - 6) % 7 == 0:
            for pl, pl_idx in ((1, 0), (-1, 1)):
                old_cnt = s.strong_pieces[pl]
                if old_cnt < 1:
                    h ^= int(_STRONG_TABLE[pl_idx, old_cnt])
                    s.strong_pieces[pl] = 1
                    h ^= int(_STRONG_TABLE[pl_idx, 1])

        s.zobrist_hash = h & 0xFFFFFFFFFFFFFFFF  # keep 64-bit unsigned
        return s

    # ── Win detection ─────────────────────────────────────────────────────────
    def _check_win(self, row, col, player):
        board = self.board
        for dr, dc in ((0, 1), (1, 0), (1, 1), (1, -1)):
            cnt = 1
            for sign in (1, -1):
                r, c = row + sign * dr, col + sign * dc
                while 0 <= r < BOARD_SIZE and 0 <= c < BOARD_SIZE \
                        and board[r, c] * player > 0:
                    cnt += 1
                    r += sign * dr
                    c += sign * dc
            if cnt >= WIN_LENGTH:
                return True
        return False

    # ── Observation ───────────────────────────────────────────────────────────
    def get_observation(self):
        p   = self.current_player
        obs = np.zeros((6, BOARD_SIZE, BOARD_SIZE), dtype=np.float32)
        obs[0] = (self.board ==  p).astype(np.float32)
        obs[1] = (self.board == 2*p).astype(np.float32)
        obs[2] = (self.board == -p).astype(np.float32)
        obs[3] = (self.board == -2*p).astype(np.float32)
        obs[4] = float(self.strong_pieces[p])
        obs[5] = float(self.strong_pieces[-p])
        return obs

    # ── Render ────────────────────────────────────────────────────────────────
    def render(self):
        SYM    = {0: ".", 1: "x", -1: "o", 2: "X", -2: "O"}
        header = "   " + " ".join(f"{c:2d}" for c in range(BOARD_SIZE))
        print(header)
        for r in range(BOARD_SIZE):
            row = f"{r:2d} " + " ".join(f" {SYM[int(self.board[r,c])]}"
                                         for c in range(BOARD_SIZE))
            print(row)
        pl = "Black" if self.current_player == 1 else "White"
        print(f"Move #{self.move_count + 1}  Player: {pl}  "
              f"Strong[B={self.strong_pieces[1]}, W={self.strong_pieces[-1]}]  "
              f"Hash={self.zobrist_hash:#018x}")
        if self.game_over:
            w = {1: "Black", -1: "White", 0: "Draw"}[self.winner]
            print(f"{w} wins!" if self.winner != 0 else "Draw!")


# ── C++ drop-in (overrides Python class if extension is built) ────────────────
try:
    from connect6s_cpp import GameState   # noqa: F811 — intentional override
    _USING_CPP = True
except ImportError:
    _USING_CPP = False

File: connect6s/game.cpp
#include "game.hpp"
#include <cassert>
#include <cstdio>
#include <random>

// ── Zobrist table initialisation ──────────────────────────────────────────────
static ZobristTables build_zobrist() {
    std::mt19937_64 rng(0xDEADBEEFCAFE0000ULL);
    ZobristTables z{};
    for (int p = 0; p < 4; ++p)
        for (int r = 0; r < BOARD_SIZE; ++r)
            for (int c = 0; c < BOARD_SIZE; ++c)
                z.cell[p][r][c] = rng();
    z.player = rng();
    for (int pi = 0; pi < 2; ++pi)
        for (int cnt = 0; cnt < 2; ++cnt)
            z.strong[pi][cnt] = rng();
    // Empty board, Black to move: XOR in "0 strong" tokens for both players
    z.init_hash = z.strong[0][0] ^ z.strong[1][0];
    return z;
}

const ZobristTables ZOBRIST = build_zobrist();

// ── GameState::clear ──────────────────────────────────────────────────────────
void GameState::clear() {
    for (auto& row : board) row.fill(EMPTY);
    current_player   = 1;
    move_count       = 0;
    strong_pieces[0] = strong_pieces[1] = 0;
    game_over        = false;
    winner           = 0;
    zobrist_hash     = ZOBRIST.init_hash;
}

// ── GameState::legal_moves ────────────────────────────────────────────────────
std::vector<Move> GameState::legal_moves() const {
    std::vector<Move> moves;
    moves.reserve(BOARD_SIZE * BOARD_SIZE * 2);
    bool has_strong = strong_for(current_player) > 0;

    for (int r = 0; r < BOARD_SIZE; ++r) {
        for (int c = 0; c < BOARD_SIZE; ++c) {
            int8_t v = board[r][c];
            if (v == EMPTY) {
                moves.push_back({r, c, false});
                if (has_strong) moves.push_back({r, c, true});
            } else if (has_strong && std::abs(static_cast<int>(v)) == 1) {
                moves.push_back({r, c, true});
            }
        }
    }
    return moves;
}

// ── GameState::apply ──────────────────────────────────────────────────────────
GameState GameState::apply(Move mv) const {
    GameState s = *this;
    int     p    = s.current_player;
    int     pidx = player_idx(p);
    int8_t  old  = s.board[mv.row][mv.col];
    int8_t  nw   = static_cast<int8_t>(mv.is_strong ? 2 * p : p);
    uint64_t& h  = s.zobrist_hash;

    // Remove old piece from hash
    if (old != EMPTY)
        h ^= ZOBRIST.cell[piece_idx(old)][mv.row][mv.col];

    // Place piece
    if (mv.is_strong) {
        assert(s.strong_pieces[pidx] > 0);
        assert(old == static_cast<int8_t>(-p));
        int old_cnt = s.strong_pieces[pidx];
        h ^= ZOBRIST.strong[pidx][old_cnt];
        s.strong_pieces[pidx]--;
        h ^= ZOBRIST.strong[pidx][s.strong_pieces[pidx]];
    } else {
        assert(old == EMPTY);
    }

    s.board[mv.row][mv.col] = nw;
    h ^= ZOBRIST.cell[piece_idx(nw)][mv.row][mv.col];

    s.move_count++;

    if (s.check_win(mv.row, mv.col, p)) {
        s.game_over = true;
        s.winner    = p;
    } else if (s.board_full()) {
        s.game_over = true;
        s.winner    = 0;
    }

    // Toggle player
    s.current_player = -p;
    h ^= ZOBRIST.player;

    s.award_strong_if_due(h);
    return s;
}

// ── GameState::check_win ──────────────────────────────────────────────────────
bool GameState::check_win(int row, int col, int player) const {
    static constexpr int dirs[4][2] = {{0,1},{1,0},{1,1},{1,-1}};
    for (auto& d : dirs) {
        int dr = d[0], dc = d[1], cnt = 1;
        for (int sign : {1, -1}) {
            int r = row + sign * dr, c = col + sign * dc;
            while (r >= 0 && r < BOARD_SIZE && c >= 0 && c < BOARD_SIZE
                   && board[r][c] * player > 0) {
                ++cnt; r += sign * dr; c += sign * dc;
            }
        }
        if (cnt >= WIN_LENGTH) return true;
    }
    return false;
}

// ── GameState::board_full ─────────────────────────────────────────────────────
bool GameState::board_full() const {
    for (auto& row : board)
        for (int8_t v : row)
            if (v == EMPTY) return false;
    return true;
}

// ── GameState::award_strong_if_due ───────────────────────────────────────────
void GameState::award_strong_if_due(uint64_t& h) {
    int next = move_count + 1;
    if (next < 6 || (next - 6) % 7 != 0) return;
    for (int pi = 0; pi < 2; ++pi) {
        if (strong_pieces[pi] < 1) {
            h ^= ZOBRIST.strong[pi][strong_pieces[pi]];
            strong_pieces[pi] = 1;
            h ^= ZOBRIST.strong[pi][1];
        }
    }
}

// ── GameState::observation ───────────────────────────────────────────────────
GameState::Obs GameState::observation() const {
    Obs obs{};
    int p    = current_player;
    int pidx = player_idx(p);
    int oidx = 1 - pidx;
    for (int r = 0; r < BOARD_SIZE; ++r)
        for (int c = 0; c < BOARD_SIZE; ++c) {
            int8_t v = board[r][c];
            obs[0][r][c] = (v ==  p)   ? 1.f : 0.f;
            obs[1][r][c] = (v == 2*p)  ? 1.f : 0.f;
            obs[2][r][c] = (v == -p)   ? 1.f : 0.f;
            obs[3][r][c] = (v == -2*p) ? 1.f : 0.f;
            obs[4][r][c] = static_cast<float>(strong_pieces[pidx]);
            obs[5][r][c] = static_cast<float>(strong_pieces[oidx]);
        }
    return obs;
}

// ── GameState::render ─────────────────────────────────────────────────────────
void GameState::render() const {
    auto sym = [](int8_t v) -> char {
        switch (v) {
            case WHITE_STRONG: return 'O';
            case WHITE_REG:    return 'o';
            case EMPTY:        return '.';
            case BLACK_REG:    return 'x';
            case BLACK_STRONG: return 'X';
            default:           return '?';
        }
    };
    std::printf("    ");
    for (int c = 0; c < BOARD_SIZE; ++c) std::printf("%2d ", c);
    std::printf("\n");
    for (int r = 0; r < BOARD_SIZE; ++r) {
        std::printf("%2d  ", r);
        for (int c = 0; c < BOARD_SIZE; ++c) std::printf(" %c ", sym(board[r][c]));
        std::printf("\n");
    }
    const char* pl = current_player == 1 ? "Black" : "White";
    std::printf("Move #%d  Player: %s  Strong[B=%d W=%d]  Hash=%016llx\n",
                move_count + 1, pl, strong_pieces[0], strong_pieces[1],
                (unsigned long long)zobrist_hash);
    if (game_over) {
        if      (winner ==  1) std::printf("Black wins!\n");
        else if (winner == -1) std::printf("White wins!\n");
        else                   std::printf("Draw!\n");
    }
}
